google/cloud/functions/internal/wrap_request.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/functions/internal/wrap_request.h"
+
+namespace google::cloud::functions_internal {
+inline namespace FUNCTIONS_FRAMEWORK_CPP_NS {
+namespace {
+class WrapBeastImpl : public google::cloud::functions::HttpRequest::Impl {
+ public:
+  using Headers = google::cloud::functions::HttpRequest::Headers;
+
+  explicit WrapBeastImpl(BeastRequest request)
+      : verb_(request.method_string()), target_(request.target()) {
+    for (auto const& h : request.base()) {
+      headers_.emplace(h.name_string(), h.value());
+    }
+    version_ = static_cast<int>(request.version());
+    // We copy all the other values before moving `body()`.
+    body_ = std::move(request).body();
+  }
+  ~WrapBeastImpl() override = default;
+
+  [[nodiscard]] std::string const& verb() const override { return verb_; }
+  [[nodiscard]] std::string const& target() const override { return target_; }
+  [[nodiscard]] std::string const& payload() const override { return body_; }
+  [[nodiscard]] Headers const& headers() const override { return headers_; }
+  static inline auto constexpr kBeastHttpVersionFactor = 10;
+  [[nodiscard]] int version_major() const override {
+    return version_ / kBeastHttpVersionFactor;
+  }
+  [[nodiscard]] int version_minor() const override {
+    return version_ % kBeastHttpVersionFactor;
+  }
+
+ private:
+  // Hold the components of the original `BeastRequest` object. We cannot hold
+  // the object directly and use its accessors because they return
+  // `std::string_view`, and we want to return `std::string const&`.
```

Comment:
Since this is C++17, do we want to return `std::string_view` too? Honest question.

Reply:
And honestly I am not sure.  I worry about `auto foo = x.thing_that_returns_string_view();`, but maybe that is an anti-pattern?

Reply:
I've always had hesitation about string_view-returning functions, for reasons like you mention, but also cases like `string_view Basename(string_view filename)`, which is incredibly easy to use wrong when the argument is a temporary (and yes, that's a real, existing function that I didn't just make up).

But I also got the impression that I'm in the minority with my caution about returning string_views. 

I say leave the const-refs as you have them for now. I added a discussion item for our team meeting so we can see if others have opinions.

Reply:
Ack.  Glad to hear it is not just me.